bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/AbstractCloudBigtableTableDoFn.java

```diff
@@ -34,7 +38,47 @@
 public abstract class AbstractCloudBigtableTableDoFn<In, Out> extends DoFn<In, Out> {
   private static final long serialVersionUID = 1L;
 
-  private static final CloudBigtableConnectionPool pool = new CloudBigtableConnectionPool();
+  static final CloudBigtableConnectionPool pool = new CloudBigtableConnectionPool();
+
+  public static void logRetriesExhaustedWithDetailsException(
+      Logger log, String context, RetriesExhaustedWithDetailsException exception) {
+
+    if (log.isDebugEnabled()) {
+      log.error("For context {}: exception occured during a bulk operation: {}", context,
+        exception.getExhaustiveDescription());
+    }
+    log.error("For context {}: exception occured during a bulk operation: {}", context,
```

Comment:
This one is very similar to the one at line#47. I don't know if we can combine them, but I am fine as it is.